### ISSUE TITLE
lookAt is backward

### ISSUE THREAD

**Issue by greggman** [toji/gl-matrix#225]
## tl;dr
- `targetTo` creates a matrix that makes something look at something else
- `lookAt` works like gluLookAt() and creates the inverse of that matrix

## Original Comment
I know this is subjective but lookAt is not generating a matrix that makes something _lookAt_ something else. It's making the inverse of that. A matrix that moves everything else such that the `eye` is at the origin. 

Why is that backward? Because in a typical scene you might have 1 camera + 50 players +100 gun turrets. You want 50 NPC heads to look the the nearest other NPC. You want the 100 turrets to target the closest NPC. That's 150 calls to lookAt except you can't use glMatrix's lookAt function for this unless you call invert on 150 matrices. There's only 1 camera so it makes more sense for `lookAt` function to make a matrix that puts the `eye` at `eye` and orients it to _lookAt_ at `target`. You then have a useful function and you can invert for that one special view matrix case when after you've made the camera _lookAt_ it's target you need view matrix for that camera to render from its point of view.

**Comment by httpdigest** [comment]
This is just how the very common GLU gluLookAt() function is defined, that has been around for more than two decades, and has been copied/adapted in various libraries for various programming languages.
If glMatrix would change its meaning, then all people knowing OpenGL/GLU and all other libraries/languages would feel uncomfortable.
You could however introduce a new function, which just uses the column vectors that lookAt also computes, and then does not transpose/invert them into row vectors but just store them in the output matrix as column vectors. And then set the translation part just to "eye" (position).
This would be quite simple to do.

**Comment by stefnotch** [comment]
How should that alternative function be named? (I would be willing to implement it.)

**Comment by andrevenancio** [comment]
I have to agree with @greggman point. Though I understand the historic reasons pointed out by @httpdigest. So regardless of opinions it would be useful to have this method in the library. `targetTo` can be an alternative. I've written a small example, but it assumes 3 vec3 are defined somewhere on the page. Might actually need to simplify the vec3 calculations and do them all inline like the method lookAt currently is doing. 

```
lookAt(out, eye, target, up) {
        vec3.subtract(zAxis, eye, target);
        vec3.normalize(zAxis, zAxis);
        vec3.cross(xAxis, up, zAxis);
        vec3.cross(yAxis, zAxis, xAxis);

        out[0] = xAxis[0];
        out[1] = xAxis[1];
        out[2] = xAxis[2];
        out[3] = 0;
        out[4] = yAxis[0];
        out[5] = yAxis[1];
        out[6] = yAxis[2];
        out[7] = 0;
        out[8] = zAxis[0];
        out[9] = zAxis[1];
        out[10] = zAxis[2];
        out[11] = 0;
        out[12] = eye[0];
        out[13] = eye[1];
        out[14] = eye[2];
        out[15] = 1;
        return out;
    }
```

**Comment by stefnotch** [comment]
Here is the optimized code + a unit test:

```
/**
 * Generates a matrix that makes something look at something else.
 *
 * @param {mat4} out mat4 frustum matrix will be written into
 * @param {vec3} eye Position of the viewer
 * @param {vec3} center Point the viewer is looking at
 * @param {vec3} up vec3 pointing up
 * @returns {mat4} out
 */
mat4.targetTo = function(out, eye, target, up) {
    var eyex = eye[0],
        eyey = eye[1],
        eyez = eye[2],
        upx = up[0],
        upy = up[1],
        upz = up[2];
        
    var z0 = eyex - target[0], 
        z1 = eyey - target[1], 
        z2 = eyez - target[2];
        
    var len = z0*z0 + z1*z1 + z2*z2;
    if (len > 0) {
        len = 1 / Math.sqrt(len);
        z0 *= len;
        z1 *= len;
        z2 *= len;
    }

    var x0 = upy * z2 - upz * z1,
        x1 = upz * z0 - upx * z2,
        x2 = upx * z1 - upy * z0;

    out[0] = x0;
    out[1] = x1;
    out[2] = x2;
    out[3] = 0;
    out[4] = z1 * x2 - z2 * x1;
    out[5] = z2 * x0 - z0 * x2;
    out[6] = z0 * x1 - z1 * x0;
    out[7] = 0;
    out[8] = z0;
    out[9] = z1;
    out[10] = z2;
    out[11] = 0;
    out[12] = eyex;
    out[13] = eyey;
    out[14] = eyez;
    out[15] = 1;
    return out;
};

```

I cannot guarantee that the unit test is correct.
```
describe("targetTo", function() {
            var eye    = new Float32Array([0, 0, 1]);
            var center = new Float32Array([0, 0, -1]);
            var up     = new Float32Array([0, 1, 0]);
            var view, up, right;

            describe("looking down", function() {
                beforeEach(function() {
                    view = new Float32Array([0, -1,  0]);
                    up   = new Float32Array([0,  0, -1]);
                    right= new Float32Array([1,  0,  0]);
                    result = mat4.targetTo(out, [0, 0, 0], view, up);
                });

                it("should transform view into local Z", function() {
                    result = vec3.transformMat4(new Float32Array(3), view, out);
                    expect(result).toBeEqualish([0, 0, 1]);
                });

                it("should transform up into local -Y", function() {
                    result = vec3.transformMat4(new Float32Array(3), up, out);
                    expect(result).toBeEqualish([0, -1, 0]);
                });

                it("should transform right into local +X", function() {
                    result = vec3.transformMat4(new Float32Array(3), right, out);
                    expect(result).toBeEqualish([1, 0, 0]);
                });

                it("should return out", function() { expect(result).toBe(out); });
            });

            describe("#74", function() {
                beforeEach(function() {
                    mat4.targetTo(out,
                        new Float32Array([0,2,0]),
                        new Float32Array([0,0.6,0]),
                        new Float32Array([0,0,-1]));
                });

                it("should transform a point 'above' into local +Y", function() {
                    result = vec3.transformMat4(new Float32Array(3), [0, 2, -1], out);
                    expect(result).toBeEqualish([0, 1, -2]);
                });

                it("should transform a point 'right of' into local +X", function() {
                    result = vec3.transformMat4(new Float32Array(3), [1, 2, 0], out);
                    expect(result).toBeEqualish([1, 2, -2]);
                });

                it("should transform a point 'in front of' into local -Z", function() {
                    result = vec3.transformMat4(new Float32Array(3), [0, 1, 0], out);
                    expect(result).toBeEqualish([0, 2, -1]);
                });
            });

            beforeEach(function() {
                eye    = new Float32Array([0, 0, 1]);
                center = new Float32Array([0, 0, -1]);
                up     = new Float32Array([0, 1, 0]);
                result = mat4.targetTo(out, eye, center, up);
            });
            it("should place values into out", function() { expect(result).toBeEqualish([
                    1, 0, 0, 0,
                    0, 1, 0, 0,
                    0, 0, 1, 0,
                    0, 0, 1, 1
                ]);
            });
            it("should return out", function() { expect(result).toBe(out); });
        });
```

I sadly cannot create a pull request since I already forked this project & added some extra stuff. (GitHub only allows 1 fork per account.)

**Comment by httpdigest** [comment]
You don't need multiple forks for that. Just `git checkout -b targetTo` (and maybe commit or stash your other changes before) and then commit onto that new branch and issue a PR for it.

**Comment by stefnotch** [comment]
@httpdigest I have never used the git command line before.

**Comment by httpdigest** [comment]
But you did get your PR done using a new branch, so: well done! :)

**Comment by stefnotch** [comment]
Yes, but as you can see https://github.com/toji/gl-matrix/pull/250 has a ton of commits that I do not wish to include. Meaning: It has a [fully fledged dual quaternion implementation](https://github.com/toji/gl-matrix/pull/231).

**Comment by stefnotch** [comment]
Ok, somehow managed to get it to work. The git command line is confusing.

**Comment by andrevenancio** [comment]
ahah :) Well is this going in? :)

**Comment by toji** [comment]
Fixed by merging #251

**Comment by stefnotch** [comment]
Reopening this issue to make it more visible.

**Comment by stefnotch** [comment]
Closing it again, now that GitHub has a built in issue duplicate finder.
